src/openstack_cloud/openstack_manager.py

```diff
@@ -434,3 +490,426 @@ def create_instance(
+        """Delete runners on Openstack.
+
+        Removes the registered runner from Github if remove_token is provided.
+
+        Args:
+            conn: The Openstack connection instance.
+            instance_names: The Openstack server names to delete.
+            remove_token: The GitHub runner remove token.
+            num_to_remove: Remove a specified number of runners. Remove all if None.
+        """
+        if num_to_remove is None:
+            num_to_remove = float("inf")
+        for instance_name in instance_names:
+            if num_to_remove < 1:
+                break
+            server: Server | None = conn.get_server(name_or_id=instance_name)
+            if not server:
+                continue
+            try:
+                self._remove_from_github(instance=server, remove_token=remove_token)
+            except GithubRunnerRemoveError as exc:
+                logger.warning("Failed to remove runner from Github %s, %s", instance_name, exc)
+
+            try:
+                if not conn.delete_server(name_or_id=instance_name, wait=True, delete_ips=True):
+                    logger.warning("Server does not exist %s", instance_name)
+                    num_to_remove -= 1
+                    continue
+            except SDKException as exc:
+                logger.error("Something wrong deleting the server %s, %s", instance_name, str(exc))
+                continue
+            num_to_remove -= 1
+
+    def reconcile(self, quantity: int) -> int:
+        """Reconcile the quantity of runners.
+
+        Args:
+            quantity: The number of intended runners.
+
+        Returns:
+            The change in number of runners.
+        """
+        github_info = self.get_github_runner_info()
+        online_runners = [runner for runner in github_info if runner.online]
+        logger.info("Found %s existing openstack runners", len(online_runners))
+
+        with _create_connection(self._cloud_config) as conn:
+            runner_by_health = self._get_openstack_runner_status(conn)
+
+            # Clean up offline (SHUTOFF) runners or unhealthy (no connection/cloud-init script)
+            # runners.
+            remove_token = self._github.get_runner_remove_token(path=self._config.path)
+            self._remove_runners(
+                conn=conn, instance_names=runner_by_health.unhealthy, remove_token=remove_token
+            )
+
+            delta = quantity - len(runner_by_health.healthy)
+
+            # Spawn new runners
+            if delta > 0:
+                self._create_runner(conn)
+            elif delta < 0:
+                self._remove_runners(
+                    conn=conn,
+                    instance_names=runner_by_health.healthy,
+                    remove_token=remove_token,
+                    num_to_remove=abs(delta),
+                )
+            else:
+                logger.info("No changes to number of runners needed")
+            return delta
+
+    def flush(self) -> int:
```

Comment:
@yanksyoon Should the `flush` method takes the `FlushMode` as argument?
The flushing needs to be done at the `charm.py` would need different flush settings.